

## 航空宇宙情報システム学第二

<h1><center> 第1回 ガイダンス・ファイル入出力 </center> </h1>

<center>

2023年4月11日

担当教員: 矢入健久

e-mail: yairi@g.ecc.u-tokyo.ac.jp
</center>




# 航空宇宙情報システム学第二ガイダンス

## はじめに

本郷（専門課程）への進学おめでとうございます。3年生のSセメスター、Aセメスターは科目数や課題が非常に多くて大変だという噂がありますが、弊科目は、**1限の授業であるということ、7号館ではなく浅野地区の情報基盤センターの演習室で行うこと以外は**、比較的「優しい」科目であると自負しています。
(2限の土屋先生の講義に十分に間に合うように、毎回**10分早く**終わるように心がけます。)

先学期(2Aセメスター)の航空宇宙情報システム学第一では、数式処理システムMathematicaや、グラフィカルなモデリング環境SIMULINKを手短に紹介した後、Pythonによるプログラミングの基本、常微分方程式の解法などを学びました。

航空宇宙情報システム学第二では、引き続きPythonを使いながら、ベクトル・行列演算、データ解析、最適化などを学んでいきたいと思います。

先学期の第一の最初の授業でも示しましたが、航空宇宙情報システム学第一(2A)、第二(3S)、第三(3A,システムコース)で扱う内容を再度書いておくと、

#### 情報システム学第一(2年Aセメ)
- モデルベーズドなシミュレーションを行うためのプログ
ラミング
- モデル≒微分方程式(数力演習と勝手に連携)

#### 情報システム学第二(3年Sセメ)
- 行列・ベクトル、データ処理
- 多変量解析(機械学習基礎)
- 数値最適化

#### 情報システム学第三(3年Aセメ)
- 状態・モデル推定(モデルとデータの融合)
- 航空・宇宙でのリアルなケーススタディ


## スケジュール (予定)

|回|日|内容|
|:-|:-|:-|
|1|4/11| ガイダンス, ファイルの読み書き |
|2|4/18| NumPyによるベクトル・行列計算 |
|3|4/25| 便利なライブラリMatploltlib, SciPy, Pandas |
|4|5/2| 線形最小二乗法と最尤法 |
|5|5/9| 正則化と変数選択 |
|6|5/16| ベイズの定理と最大事後確率推定 |
|7|5/23| 線形判別（ロジスティック回帰） |
|8|6/6 (5/30)*| 主成分分析と多次元尺度法 |
|9|6/13| クラスタリング（k-means法, 混合ガウス分布モデル) |
|10|6/20|探索（経路探索, 幅優先, 深さ優先, Dijkstra, A* アルゴリズム）|
|11|6/27|制約なし最適化 |
|12|7/4|非線形最小二乗法 |
|13|7/11|制約あり最適化 |

(*) 8回目は今の所6/6を予定していますが、学会出張が入る可能性があり、その場合は5/30の補講日を使わせてもらうつもりです。

今年新たな試みとして、機械学習の基本である多変量解析の話題（線形判別、主成分分析、クラスタリング、正則化、など）を扱ってみたいと考えています。

# 諸連絡
* 第一では駒場の情報教育棟の演習室で行いましたが、第二は本郷(浅野地区)の情報基盤センターの演習室で行います。7号館から少し歩きますが、気分転換だと思ってください。。
* 演習室に設置されているECCSの端末にログインするためには、UTokyoアカウントと連動しているECCSアカウントが必要です。ユーザー名は10桁の番号、パスワードはUTokyoアカウントと同じ（はず）です。
* 授業ポータルサイトへのアクセスには<font color=red>ECCSクラウドメールにサインインしておく</font>必要があります。
* 成績は（１）出席（２）小課題、に基づいて判定します。
* 小課題は隔週で6回程度出題することを予定しています。
* 矢入研M1の中島琢登君がティーチングアシスタント(TA)を務めてくれます。
  * 小課題の採点、質問の回答、などを担当します。
  * とりあえず、質問や相談事が有ったら、気軽にポータルサイトのコメント欄で知らせてください。メールでも構いません。
* 演習室の端末を使って講義を受けることを推奨しますが、自前のノートPCを持参してUTokyo WIFIに接続して使ってもらっても構いません。



---

# ファイル入出力

## はじめに
先学期、Pythonによるプログラミングの基本を一通りやりましたが、時間が足りなくて扱わなかったことの一つが、「ファイルを開いてデータを読み書きする」ことでした（その代わり、Webページからデータをスクレイピングしてみましたが。）。

前述のように、今学期はデータ処理・分析を一つのテーマとしているので、プログラム(ソースコード)からデータファイルを読み込むのは必須です。そこで、今日はテキストファイルをデータとして読み込んで処理する基本的な方法を扱います。

## 今日扱うデータファイル
* <font color="red">**weather_tokyo_1980_2023.csv**</font> :
1980年6月1日から2023年4月7日まで約43年間の東京の気象情報（年月日、最高気温、最低気温、降雨量、日照時間、平均風速、平均蒸気圧、平均雲量）を<font color="blue">**カンマ区切り**</font>で列挙したものです。
* [気象庁「過去の気象データ・ダウンロード」](http://www.data.jma.go.jp/gmd/risk/obsdl/) よりダウンロードしたデータから作成しました。
* 先頭行（ヘッダー）には、項目名が書かれていて、2行目以降に日付と数値データが並んでいます。
* <font color="blue">テキストエディタで</font>開いてみると、
```
date,temp_hi,temp_lo,rain,sun,wind,vapor,cloud
1980/6/1,25,18.7,0,1.1,1.7,20.1,10
1980/6/2,27.3,21.1,13.5,5.4,3.2,21.9,10
1980/6/3,28.3,20.4,2,5.7,2.4,22.3,9
:
:
2021/6/8,31.4,18.8,0,9,2.2,19.6,6.8
2021/6/9,31.1,19.5,0,13.1,2.8,16.4,1.5
2021/6/10,30.7,19,0,13.2,2.7,18.7,0.8
```
* このようなカンマ区切りのテキストで表現されたデータファイルは、<font color="blue">csv</font>フォーマットと呼ばれます。
* Excelなどのスプレッドシートアプリで開くと、テーブル状に表示されます。

## Colaboratory でファイルを開く準備

ColabはブラウザさえあればPythonプログラミングができて非常に便利なのですが、クラウド上で動作しているので、データファイルにアクセスするのが少々面倒です。

おそらく、一番簡単と思われるのは、
1. 使いたいファイルをGoogleドライブにアップロードする。
2. GoogleドライブをColabからマウントする。
3. Colab ノートブックでアップロードしたファイルを開く。

という手順なので、ここではこれに従います。

必須ではないのですが、まず、Colab上の現在のフォルダ(ディレクトリ)を確かめておきます。

In [ ]:
import os
# カレントディレクトリ
os.getcwd()

'/content'

### データファイルをローカルにダウンロード

とりあえず、[weather_tokyo_1980_2023.csv](https://drive.google.com/file/d/1jjwQ3TtFUavc05BTygKiQEYsxkoOq1AG/view?usp=share_link) を、現在作業しているPCのローカルフォルダ(*注)にダウンロードしましょう。フォルダはどこでも構いません。

  - (*注) ECCSの端末の場合、ホームフォルダがサーバー上にあるので厳密には「ローカル」とは呼べませんが、ここでは気にしないことにしましょう。

### Googleドライブにアップロード

ローカルにダウンロードしたファイル(weather_tokyo_1980_2023.csv)を、自身のGoogleドライブの<font color="red">"マイドライブ"直下</font>にアップロードしてください。
* <font color="blue">ファイルやディレクトリの「パス」について理解している</font>人は、
”マイドライブ”以外のフォルダ（サブフォルダ）にファイルを置いても全く問題ありません。

### Googleドライブをマウント(*)


* Colabの左側にあるフォルダアイコンをクリックします。

![](https://drive.google.com/uc?export=view&id=1BvF_JtsyT7cbwJliubNea56dGpJTpAzj)

* GOOGLEドライブのアイコンをクリックします。

![](https://drive.google.com/uc?export=view&id=17zMWuO1ouKWFLhhqPGDhxaFn-TiJY4sn)

* 「アクセスを許可しますか」のような確認があるので、「接続する」を選択

![](https://drive.google.com/uc?export=view&id=1_4p9oWDaTC7Ww-0yWhVmL6RXxoIcgzqK)

* マウントされた(アクセス可能になった)Googleドライブが、"drive"という名前で現れます。

![](https://drive.google.com/uc?export=view&id=18-yo4PGx8IHOmuIsR7lYGNXA0h2eafR-)

これで、ColabからGoogleドライブ内のファイルにアクセスできるようになります。

* (雑談) 「マウント」という言葉に「？」と思う人も多いかと思います。計算機(コンピュータ)業界では、「現在使っている処理系に、別のファイルシステムの接続して利用可能にすること」を「マウント」と呼びます。例えば、USBメモリをPCのUSB端子に挿すと、USBメモリ上のファイルやフォルダにアクセスできるようになりますよね。この状態を「マウント」と呼びます。



### ファイルが存在するかを確認

ここまでの作業が想定通りに出来ていれば、データファイル(weather_tokyo_1980_2023.csv)に、Colab上で、"drive/My Drive/weather_tokyo_1980_2023.csv"というパスでアクセスできるはずです。

以下のコードを実行してみてください。

In [ ]:
import os.path
os.path.exists("drive/My Drive/weather_tokyo_1980_2023.csv")

True

結果が、<font color="blue">"True"となっていれば、OK</font>です。

<font color="red">"False"となった場合は、どこかが違っている</font>ので、冷静になってもう一度やり直してください。

## CSVファイルの読み込み

* 本当は、CSVファイルを扱う場合、**csvモジュール**を使った方が確実です。また、より汎用かつ強力なデータ分析ライブラリとして、**後日紹介するpandas**があります。ですが、<font color="red">ここでは練習を兼ねて、組み込み関数`open`や文字列メソッドを使った基本的な方法で読み込みます。</font>
* まず、カンマのことは気にせずに、全行をリストに読み込んで、読み込まれた行数を表示してみます。
```
>>> lines = open('drive/My Drive/weather_tokyo_1980_2023.csv').readlines()
>>> print(len(lines)) # 読み込んだ行数
```
* 先頭行（ヘッダー）を表示させてみると、
```
>>> print(lines[0])
date,temp_hi,temp_lo,rain,sun,wind,vapor,cloud
```

In [ ]:
fh = open('drive/My Drive/weather_tokyo_1980_2023.csv')
line = fh.readline()
print(line)

date,temp_hi,temp_lo,rain,sun,wind,vapor,cloud



In [ ]:
line = fh.readline()
print(line)

1980/6/1,25,18.7,0,1.1,1.7,20.1,10



In [ ]:
for i in range(10):
  line = fh.readline()
  print(line)

1980/6/2,27.3,21.1,13.5,5.4,3.2,21.9,10

1980/6/3,28.3,20.4,2,5.7,2.4,22.3,9

1980/6/4,24.9,20.4,0,5.7,3.4,14.6,7

1980/6/5,27.2,19.4,0,11,2.2,16.5,5.3

1980/6/6,28.5,19.5,0,11,2.6,18.6,3.8

1980/6/7,28.2,19.4,0,12.6,3.8,20.3,3.5

1980/6/8,24.9,20.4,12,0,4,22.8,9.8

1980/6/9,24.8,22,11,0,6.3,24,8.8

1980/6/10,27.2,22.7,0,0.5,2.6,24.1,10

1980/6/11,25.7,21.1,0,3.6,2.8,18.1,9.5



In [ ]:
lines = open('drive/My Drive/weather_tokyo_1980_2023.csv').readlines() # 全行読み込み
print(len(lines)) # 行数
print(lines[0]) # 先頭行

15652
date,temp_hi,temp_lo,rain,sun,wind,vapor,cloud



**(練習)** 先頭行(ヘッダー)`lines[0]='date,temp_hi,temp_lo,rain,sun,wind,vapor,cloud'` を、カンマ(`,`)で区切って、リスト
``
['date', 'temp_hi', 'temp_lo', 'rain', 'sun', 'wind', 'vapor', 'cloud']
``
を作ってください。

In [ ]:
lines[0].strip().split(',')

['date', 'temp_hi', 'temp_lo', 'rain', 'sun', 'wind', 'vapor', 'cloud']

**(練習)** 先頭行の次の行も同様にカンマで区切った文字列のリストにしてください。

In [ ]:
lines[1].strip().split(',')

['1980/6/1', '25', '18.7', '0', '1.1', '1.7', '20.1', '10']

## データの整形
* `lines`リストの各要素`lines[i]`には、ファイルの各行が文字列として入っています。末尾には改行文字'\n'が入っているので、`strip`メソッドで除去し、さらに、<font color="blue"> `split`メソッドに区切り文字カンマ(`,`)を引数として渡せば、行を分解する</font>ことができます。
```
>>> print(lines[1].strip().split(','))
['1980/6/1', '25', '18.7', '0', '1.1', '1.7', '20.1', '10']
```
* データをどのようにオブジェクト化するかは、何通りか考えられますが、<font color="blue">今回は、辞書型オブジェクト`d`のキーを各項目（列名）にし、値にその列のデータをリストで持つようにします。</font>例えば、`d['date'][0]` は'1980/6/1'になります。
* まず、空の辞書`d`を作り、先頭行`lines[0]`から読み込んだ各項目名をキーとし、値に空のリストを割当ます。
```
d = dict() # 空の辞書
columns = lines[0].strip().split(',') # 項目名のリスト
for k in columns:
    d[k] = [] # 各項目名をキーとして空のリストを初期値にする
```
* 次の行からは、各リストに値を追加していきます。
  * 今回は'date'項目は文字列のまま扱い、他の項目は数値（実数値）に型変換しましょう。少々長くなりますが、
```
for line in lines[1:]: # 2行目から最終行まで
      lst = line.strip().split(',') # カンマ区切りでリストに
      for i in range(len(lst)): # 列の数だけループ
        k = columns[i] # i列目の項目名
        if k == 'date':  # 日付列のときは文字列のまま追加
          d[k].append(lst[i])
        else: # それ以外の列は実数値に変換して追加
          d[k].append(float(lst[i]))
```

* 例えば、最高気温のデータは、`d['temp_hi']`でアクセスできます。
```
print(d['temp_hi'])
```
* この方式では、列単位でデータをリスト化しているので、行単位で参照する（例えば、'2020/06/15'の日照時間を参照する）のには不便ですが、今日は目をつぶることにします。
  * Pythonで本格的にデータ処理を行う場合には、pandasライブラリを使うのがお薦めなので、後日その方法を紹介します。

In [ ]:
d = dict() # 空の辞書を作成
columns = lines[0].strip().split(',') #先頭行から項目リストを取得
for k in columns: # 各項目をキーとして、空のリストを割当
  d[k] = []

In [ ]:
print(d)

{'date': [], 'temp_hi': [], 'temp_lo': [], 'rain': [], 'sun': [], 'wind': [], 'vapor': [], 'cloud': []}


In [ ]:
for line in lines[1:]: # 2行目から最終行まで
  lst = line.strip().split(',') # カンマ区切りでリストに
  for i in range(len(lst)): # 列の数だけループ
    k = columns[i] # i列目の項目名
    if k == 'date':  # 日付列のときは文字列のまま追加
      d[k].append(lst[i])
    else: # それ以外の列は実数値に変換して追加
      d[k].append(float(lst[i]))

In [ ]:
d['rain'][1000:1010]

[0.0, 0.0, 0.0, 0.0, 0.0, 6.0, 0.0, 0.5, 0.0, 0.0]

## 読み込んだデータを使ってみる
早速、辞書型オブジェクトに格納された天気データを使ってみましょう。

* `d['項目名'][インデクス]` で各データ要素にアクセスできます。
* 例： 10000日目の日付とその日の最高気温
```
idx = 10000
print(d['date'][idx],d['temp_hi'][idx])
```

In [ ]:
idx = 10000
print(d['date'][idx],d['temp_hi'][idx],d['rain'][idx])

2007/10/17 21.7 0.0



**(練習)** 約43年間における東京の**最高気温(temp_hi)の最高値と平均**を求めてください。
リストに対する組み込み関数 `max()`,`sum()`,`len()`を使うと良いと思います。

In [ ]:
# 最高気温の最高
max(d['temp_hi'])

39.5

In [ ]:
# 最高気温の平均
sum(d['temp_hi'])/len(d['temp_hi'])

20.22954443805518

ついでなので、上の練習問題と同様に、**最低気温(temp_lo)**の平均を求めてみてください。期待に反して、`nan` という結果が表示されると思います。

In [ ]:
# 最低気温の最低は求まるが,
min(d['temp_lo'])

-4.0

In [ ]:
# 最低気温の平均はnanになってしまう
sum(d['temp_lo'])/len(d['temp_lo'])

nan

## NaN ("not a number")

ここで出力された `nan` というのは、"not a number" の略、すなわち、非数を表す値です。本来の数値であるべきなのに、何らかの理由で具体的な数値にならなかった場合に用いる値です。

組み込み関数 `sum()`は、引数として、整数や実数のリストが与えられたときにその総和を返しますが、リストの中に1つでも `nan`が含まれると、総和を計算することができないので、`nan`を返します。`sum()`に限らず、`nan`に何らかの数値を足したり引いたりしても`nan`になります。

In [ ]:
1*2*float('nan')

nan

`nan`かどうかを判定するには、`math`モジュールに含まれる`isnan()`関数を使います。

In [ ]:
import math
print(math.isnan(1.0)) # 普通の数値なので当然False
print(math.isnan(float('nan'))) #nan

False
True


先程の例題で、どうして「(日別)最低気温の平均値」が`nan`になってしまったのか、原因を探ってみましょう。

一番有り得そうなのは、リスト`d["temp_lo"]`の要素の中に`nan`が含まれていたから、ということです。これを確かめてみましょう。

In [ ]:
for i in range(len(d['temp_lo'])):
  if math.isnan(d['temp_lo'][i]):
    print(i,d['date'][i],d['temp_lo'][i])

5160 1994/7/18 nan


確かに、リスト`d['temp_lo`]の5160番目要素が`nan`になっていました。つまり、1994年7月18日は何らかの理由で東京の最低気温を記録することができなかった、ということのようです。

**(雑談)** 気になったのでネット検索したり、ChatGPTに聞いてみたり（笑）しましたが、どうしてこの日の東京の最低気温が残っていないのかわかりませんでした。そこで、東大図書館が契約している新聞データベースでこの日の毎日新聞を調べてみたところ、

```
「関東で局地雷雨　ＪＲダイヤ混乱、道路冠水 1994.07.19　東京朝刊　27頁　社会　(全808字)　」

関東地方は十八日午後、上空に寒気が入り込んだ影響で大気が不安定になり、局地的な
雷を伴った大雨に見舞われた。東京・大手町では一時間に四六ミリの雨量を記録。
この雨で、ＪＲのダイヤが大きく乱れたほか、道路冠水などが相次いだ。
(中略)
各地の雨量は、大手町が午後二時から三時までに四六ミリを記録したほか、
(中略)
この雷雨で、東京、神奈川、埼玉の三都県で計十六万八千五百世帯が一時停電した。
```
という記事がありました。もしかしたら、このゲリラ豪雨が関係しているのかもしれませんね。

どうでも良い話ですが、遠い記憶を辿ると、、1994年7月18日頃は、わたしは修士課程1年で、所属していた田辺・堀・中須賀研究室の夏のイベント、中間発表会と夏合宿の準備をしていたと思います。

**(練習)** `nan`となっている日を除いて、(日別)最低気温の平均値を求めてください。

In [ ]:
# for文で一つ一つチェックしながら、合計と日数を更新する方法。
tot,days = 0,0
for tmp in d['temp_lo']:
  if not math.isnan(tmp):
    tot,days = tot+tmp,days+1
# 最低気温の平均
tot /days

12.965936102236373

In [ ]:
# リスト内包表記でフィルター処理
lst_temp_lo = [tmp for tmp in d['temp_lo'] if not math.isnan(tmp)]
sum(lst_temp_lo)/len(lst_temp_lo)

12.965936102236373

## 過去最高○○または最低○○の日を知るには

過去43年間での東京の最高気温、最低気温、さらには、(1日の)最大降水量は、それぞれ、
```
print("最高気温:",max(d['temp_hi']))
print("最低気温:",min(d['temp_lo']))
print("最大降水量",max(d['rain']))
```
のように、組み込み関数`max()`,`min()`を使えば容易に知ることができます。

In [ ]:
print("最高気温:",max(d['temp_hi']))
print("最低気温:",min(d['temp_lo']))
print("最大降水量",max(d['rain']))

最高気温: 39.5
最低気温: -4.0
最大降水量 259.5


しかし、これでは最高気温や最低気温、最大降水量が、何年何月何日だったのかわかりません。
後日扱うnumpyやpandasを使えば、それらを比較的簡単に求めることができるのですが、リスト型だと以下の手順を踏む必要があり少々面倒です。
1. `max()`関数を使ってリストの中から最大値（または最小値）を求める。
1. リスト型の`index()`メソッドを使って、上で求めた値を取る要素のインデクスを求める。
1. そのインデクスを使って、日付を求める。

例えば、東京で最高気温を記録した年月日を求めるには、次のようにします。

In [ ]:
highest_temp = max(d['temp_hi'])
idx = d['temp_hi'].index(highest_temp)
date_highest_temp = d['date'][idx]
print("東京の最高気温",highest_temp,"を記録したのは、",date_highest_temp,"です。")

東京の最高気温 39.5 を記録したのは、 2004/7/20 です。


**(練習)** 東京の過去43年で、最低気温を記録した日、(1日の)最大降水量を記録した日を調べてください。

In [ ]:
temp_lowest = min(d['temp_lo'])
idx = d['temp_lo'].index(temp_lowest)
print("東京の最低気温",temp_lowest,"を記録したのは、",d['date'][idx],"です。")

東京の最低気温 -4.0 を記録したのは、 2018/1/25 です。


## 練習問題いろいろ

読み込んだ東京の天気データを使って、色々調べてみましょう。

**(練習問題)** 1日に200mm以上の雨が降った日の日付と降水量を表示してください。
* `d['rain']`のリストを最初から最後まで走査しながら、値が200以上のときの日付と降雨量を表示すれば良いです。
* `zip()`関数を使って、各日の日付と降雨量をタプルで受け取って処理すると少しエレガントです。


**(練習問題)** 寒暖差が最も大きかった日の日付とその日の最高・最低気温を求めてください。
* 基本的な考え方: 1日ずつ最高気温と最低気温の差を取り、それまでの記録よりも大きければ更新する。
* まず気温差のリストを作り、その最大値のインデクスを求めるというやり方も良いと思います。


**(練習問題)** 最高気温が35度以上だった日(猛暑日)の数を年別に求めてください。
* ヒント: 日付の文字列が、'YYYY/MM/DD' のような形式になっているので、西暦年('YYYY')の部分を取り出し、それをキーとする新たな辞書型で猛暑日の日数をカウントします。

**(練習問題)** 降水量が最も多かった１週間（連続する7日間）を調べてください。

**(練習問題)** 東京で43年間で最も雨の降らない月日、最も雨の降る月日のトップ10を求めてください。

**(練習問題)** 日別最高気温の年平均の変遷を調べてください。

この練習問題をやってみると、<font color="red">40年間で2度くらい最高気温平均が上昇しているという衝撃の結果</font>が出ると思います。ただ、これが全て地球温暖化の影響かというとそういうわけではなく、都市部のヒートアイランド現象による部分が大きいみたいです。

**(練習問題)** 最低気温の月平均の変遷を調べてください。



## ファイルへの書き込み

* ファイルを<font color="red">書き込みモードで</font>開いて、文字列を書き込みます。
```
>>> fout = open(‘output.txt’,‘w’) # 書き込みモード
>>> fout.write(‘This is the first line.\n’) # writeメソッド
>>> line = 'This is the last line. \n‘
>>> fout.write(line)
>>> fout.close() # 必ず閉じよう
```
* `writelines`メソッドを使って、文字列のリストをまとめて書き込む方法もあります。
```
>>> lines = ['First line\n','Second line\n','Last line']
>>> open('output.txt','w').writelines(lines)
```


## 数値の書き込みと書式演算子・`format`メソッド

* 数値をテキストファイルに書き出す場合は、文字列に変換する必要があります。例えば、変数`x`,`y`,`z`に代入されている数値を、カンマ区切りで文字列にして`output.csv`に出力したい場合、
* (方法1) 数値を`str`関数で文字列に変換して連結する（お手軽だけど、細かい調整は難しい）
```
fout = open('output.csv','w')
x,y,z = 2.5, 10.9, 22.3 #タプル代入
line = str(x)+','+str(y)+','+str(z)+'\n'
fout.write(line)
fout.close() # 使い終わったファイルは閉じる
```
  * 多数の数値が含まれているリストをカンマ区切りの文字列に変換する場合は、`join`メソッドを使うと便利です。
  ```
  line = ','.join([str(x) for x in lst])
  ```
* (方法2) `%`演算子で文字列の中に数値を埋め込む(C言語的。古い。)
```
line = '%f,%f,%f\n' % (x,y,z)
```
* (方法3) `format`メソッドで文字列に埋め込む(新しい方法)
```
line = '{},{},{}\n'.format(x,y,z)
```
* `format`メソッドはきめ細かく出力を調整することができます。[Python公式ドキュメント](https://docs.python.org/ja/3/tutorial/inputoutput.html)などを読んでください。



In [ ]:
# (方法1) 数値をstr関数で文字列に変換して連結
fout = open('output.csv','w')
x,y,z = 2.5,10.9,22.3
line = str(x)+','+str(y)+','+str(z)+'\n'
fout.write(line)
fout.close()

In [ ]:
# (方法2) %演算子で文字列の中に数値を埋め込む
fout = open('output.csv','w')
line = '%f,%f,%f' % (x,y,z)
fout.write(line)
fout.close()

In [ ]:
# (方法3) `format`メソッドで文字列に埋め込む
fout = open('output.csv','w')
line = '{},{},{}\n'.format(x,y,z)
fout.write(line)
fout.close()